framework/include/executioners/FixedPointSolve.h

```diff
@@ -0,0 +1,254 @@
+
+  static InputParameters validParams();
+
+  /**
+   * Iteratively solves the FEProblem.
+   * @return True if solver is converged.
+   */
+  virtual bool solve() override;
+
+  /// Enumeration for fixed point convergence reasons
+  enum class MooseFixedPointConvergenceReason
+  {
+    UNSOLVED = 0,
+    CONVERGED_NONLINEAR = 1,
+    CONVERGED_ABS = 2,
+    CONVERGED_RELATIVE = 3,
+    CONVERGED_CUSTOM = 4,
+    REACH_MAX_ITS = 5,
+    DIVERGED_MAX_ITS = -1,
+    DIVERGED_NONLINEAR = -2,
+    DIVERGED_FAILED_MULTIAPP = -3
+  };
+
+  /**
+   * Get the number of fixed point iterations performed
+   * Because this returns the number of fixed point iterations, rather than the current
+   * iteration count (which starts at 0), increment by 1.
+   *
+   * @return Number of fixed point iterations performed
+   */
+  unsigned int numFixedPointIts() const { return _fixed_point_it + 1; }
+
+  /// Deprecated getter for the number of fixed point iterations
+  unsigned int numPicardIts() const
+  {
+    mooseDeprecated("numPicards() is deprecated. Please use numFixedPointIts() instead.");
+
+    return _fixed_point_it + 1;
+  }
+
+  /// Check the solver status
+  MooseFixedPointConvergenceReason checkConvergence() const { return _fixed_point_status; }
+
+  /// This function checks the _xfem_repeat_step flag set by solve.
+  bool XFEMRepeatStep() const { return _xfem_repeat_step; }
+
+  /// Clear fixed point status
+  void clearFixedPointStatus() { _fixed_point_status = MooseFixedPointConvergenceReason::UNSOLVED; }
+
+  /// Whether or not this has fixed point iterations
+  bool hasFixedPointIteration() { return _has_fixed_point_its; }
+
+  /// Set relaxation factor for the current solve as a SubApp
+  void setMultiAppRelaxationFactor(Real factor) { _secondary_relaxation_factor = factor; }
+
+  /// Set relaxation variables for the current solve as a SubApp
+  void setMultiAppTransformedVariables(const std::vector<std::string> & vars)
+  {
+    _secondary_transformed_variables = vars;
+  }
+
+  /// Set relaxation postprocessors for the current solve as a SubApp
+  virtual void setMultiAppTransformedPostprocessors(const std::vector<std::string> & pps)
+  {
+    _secondary_transformed_pps = pps;
+  }
+
+  /// Allocate storage for secondary transformed stuff
+  virtual void allocateStorageForSecondaryTransformed() = 0;
+
+  /// Whether sub-applications are automatically advanced no matter what happens during their solves
+  bool autoAdvance() const;
+
+  /// Mark the current solve as failed due to external conditions
+  void failStep() { _fail_step = true; }
+
+protected:
+  /// Save the variable values as a SubApp
+  virtual void savePreviousVariableValuesAsSubApp() = 0;
+
+  /// Save the postprocessor values as a SubApp
+  virtual void savePreviousPostprocessorValuesAsSubApp() = 0;
+
+  /// Whether to use the fixed point algorithm (relaxed Picard, Secant, ...) instead of Picard
+  virtual bool useFixedPointAlgorithmUpdate(bool as_main_app) = 0;
```

Comment:
Isn't picard itself a fixed point algorithm though (you say use picard instead of picard here :P ?  I think this needs a different/better name and doc comment.  Like solutionUpdateReady - or something along those lines.

Reply:
Picard is the default algorithm. Relaxed Picard isnt quite Picard. 
updating the postprocessors isnt really updating the solution though. It's more often updating the boundary conditions. 
Also 'ready' isnt that clear for Steffensen where you purposefully skip a step

Reply:
Yeah - my suggestion wasn't really a good one either.  Maybe `fallbackToPicard` might be better...  The doc comment could certainly be better as well.  All these (pure) virtual functions could probably use some good doc comments and names so when people come along later to modify or implement new algos, they won't just be blindly copy-pasting - or tracking you down forcing you to remember why you did something a certain way :P .